### PR TITLE
Samd5e5 progmem

### PR DESCRIPTION
**Pull request by leomar95**
## Summary
arch: samd5e5:
- Add MTD progmem support.
- NVM USER PAGE IOCTLs support.

boards: metro-m4 Add support for:
- SmartFS inicialization.
- AT24 EEPROM.
- GPIO dev support.
- BQ27426 gauge inicialization.

drives: power

- Add BQ27426 fuel gauge support.

## Impact
New features available.

## Testing
Build check only.

Signed-off-by: Leomar Mateus Radke leomar@falker.com.br

### CODE REVIEW

**Comment by acassis** [comment]
It still comming with issues, now it reduced from 51 to 21

I do this way on github:

https://acassis.wordpress.com/2020/03/31/syncing-my-github-fork/

**Comment by v01d** [comment]
To get rid of the merge commits you could try squashing. An alternative is doing it manually by cherry picking your changes.

**Comment by leomar95** [comment]
> To get rid of the merge commits you could try squashing. An alternative is doing it manually by cherry picking your changes.

Thanks for your help @v01d . I removed the merge commits, it's worked!

File: boards/arm/samd5e5/metro-m4/src/metro-m4.h [hunk 1]
```diff
@@ -71,6 +71,46 @@
 #define PORT_RED_LED (PORT_OUTPUT | PORT_PULL_NONE | PORT_OUTPUT_SET | \
                       PORTA | PORT_PIN16)
 
+#define PORT_D8	  (PORT_OUTPUT     | PORT_PULL_NONE | PORT_OUTPUT_SET | PORTA | PORT_PIN21)
```
Comment (acassis) [inline boards/arm/samd5e5/metro-m4/src/metro-m4.h:74]: Please, try to align this line

File: boards/arm/samd5e5/metro-m4/src/sam_bq27426.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
+/****************************************************************************
+ * boards/arm/samd5e5/metro-m4/src/sam_bq27426.c
+ *
+ *   Copyright 2020 Falker Automacao Agricola LTDA.
```
Comment (acassis) [inline boards/arm/samd5e5/metro-m4/src/sam_bq27426.c:4]: We cannot use our Copyright with Apache Foundation (too bad... I also liked the BSD way)

File: boards/arm/samd5e5/metro-m4/src/sam_bringup.c [hunk 3]
```diff
@@ -137,12 +142,47 @@ int sam_bringup(void)
...
 #ifdef CONFIG_FS_AUTOMOUNTER
   /* Initialize the auto-mounter */
 
   sam_automount_initialize();
 #endif
 
+#ifdef CONFIG_BQ27426
+  /* Configure and initialize the BQ2426 distance sensor */
+
+  ret = sam_bq27426_initialize("/dev/batt1");
```
Comment (acassis) [inline boards/arm/samd5e5/metro-m4/src/sam_bringup.c:175]: "Starting From Zero Got Nothing to Lose" ;-)

File: boards/arm/samd5e5/metro-m4/src/sam_gpio.c [hunk 4]
```diff
@@ -0,0 +1,341 @@
+/****************************************************************************
+ * boards/arm/samd5e5/metro-m4/src/sam_gpio.c
+ *
+ *   Copyright 2020 Falker Automacao Agricola LTDA.
+ *   Author: Leomar Mateus Radke <leomar@falker.com.br>
```
Comment (acassis) [inline boards/arm/samd5e5/metro-m4/src/sam_gpio.c:5]: Ditto

File: boards/arm/samd5e5/metro-m4/src/sam_smartfs.c [hunk 5]
```diff
@@ -0,0 +1,238 @@
+/****************************************************************************
+ * boards/arm/samd5e5/metro-m4/src/sam_smartfs.c
+ *
+ *   Copyright 2020 Falker Automacao Agricola LTDA.
```
Comment (acassis) [inline boards/arm/samd5e5/metro-m4/src/sam_smartfs.c:4]: Ditto

File: include/nuttx/power/bq27426.h [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+/* Configuration Class, Subclass ID's */
+
+/*  To access a subclass of the extended data, set the DataClass()
+ *  function with one of these values.
+ */
+
+/* Configuration Classes */
+
+#define BQ27426_ID_SAFETY            2  /* Safety */
```
Comment (acassis) [inline include/nuttx/power/bq27426.h:181]: Please, try to align those value to start all in the same column, this block and all others

File: arch/arm/src/samd5e5/sam_progmem.c [hunk 7]
```diff
@@ -0,0 +1,1015 @@
+/****************************************************************************
+ * arch/arm/src/samd5e5/sam_progmem.c
+ *
+ *   Copyright 2020 Falker Automacao Agricola LTDA.
+ *   Author: Leomar Mateus Radke <leomar@falker.com.br>
+ *   Author: Ricardo Wartchow <wartchow@gmail.com>
```
Comment (xiaoxiang781216) [inline arch/arm/src/samd5e5/sam_progmem.c:6]: The new apache copyright don't allow to add the contributor name.

File: arch/arm/src/samd5e5/Make.defs [hunk 8]
```diff
@@ -177,3 +177,7 @@ endif
 ifeq ($(CONFIG_SAMD5E5_RTC),y)
 CHIP_CSRCS += sam_rtc.c
 endif
+
+ifeq ($(CONFIG_SAMD5E5_PROGMEM),y)
+CHIP_CSRCS += sam_progmem.c
+endif
```
Comment (xiaoxiang781216) [inline arch/arm/src/samd5e5/Make.defs:183]: add one newline

File: arch/arm/src/samd5e5/sam_progmem.h [hunk 9]
```diff
@@ -0,0 +1,59 @@
+/****************************************************************************
+ * arch/arm/src/samd5e5/sam_progmem.h
+ *
+ *   Copyright 2020 Falker Automacao Agricola LTDA.
+ *   Author: Leomar Mateus Radke <leomar@falker.com.br>
+ *   Author: Ricardo Wartchow <wartchow@gmail.com>
```
Comment (xiaoxiang781216) [inline arch/arm/src/samd5e5/sam_progmem.h:6]: remove author.

File: boards/arm/samd5e5/metro-m4/src/sam_at24.c [hunk 10]
```diff
@@ -0,0 +1,129 @@
+/****************************************************************************
+ * boards/arm/samd5e5/metro-m4/src/sam_at24.c
+ *
+ *   Author: Leomar Mateus Radke <leomar@falker.com.br>
+ *   Author: Ricardo Wartchow <wartchow@gmail.com>
```
Comment (xiaoxiang781216) [inline boards/arm/samd5e5/metro-m4/src/sam_at24.c:5]: remove the author.

File: boards/arm/samd5e5/metro-m4/src/sam_bq27426.c [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+{
+    FAR struct battery_gauge_dev_s *bq27426_m4;
+
+    bq27426_m4 = (FAR struct battery_gauge_dev_s *)bq27426_initialize(
+                                                   g_i2c5_dev,
+                                                   BQ27426_I2C_ADDRESS,
+                                                   100000);
+  return battery_gauge_register(devname, bq27426_m4);
+}
+#endif /* CONFIG_BQ27426 */
```
Comment (xiaoxiang781216) [inline boards/arm/samd5e5/metro-m4/src/sam_bq27426.c:77]: add new line

File: boards/arm/samd5e5/metro-m4/src/sam_smartfs.c [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+    {
+      syslog(LOG_ERR, "ERROR: SmartFS initialization failed: %d\n", ret);
+      return ret;
+    }
+
+    #endif
+
+  return OK;
+}
+#endif /* CONFIG_FS_SMARTFS */
```
Comment (xiaoxiang781216) [inline boards/arm/samd5e5/metro-m4/src/sam_smartfs.c:237]: add newline

File: drivers/power/bq27426.c [hunk 13]
```diff
@@ -0,0 +1,460 @@
+/****************************************************************************
+ * drivers/power/bq27426.c
+ *
+ * Lower half driver for BQ27426 battery fuel gauge
+ *
+ *   Author: Leomar Mateus Radke <leomar@falker.com.br>
+ *   Author: Ricardo Wartchow <wartchow@gmail.com>
```
Comment (xiaoxiang781216) [inline drivers/power/bq27426.c:7]: remove

File: include/nuttx/power/bq27426.h [hunk 14]
```diff
@@ -0,0 +1,213 @@
+/****************************************************************************
+ * include/nuttx/power/bq27426.h
+ * Lower half driver for bq27426 battery fuel gauge.
+ *
+ *   Author: Leomar Mateus Radke <leomar@falker.com.br>
+ *   Author: Ricardo Wartchow <wartchow@gmail.com>
```
Comment (xiaoxiang781216) [inline include/nuttx/power/bq27426.h:6]: remove

**Comment by acassis** [comment]
Hi @leomar95 and @xiaoxiang781216 can we merge it now?

**Comment by xiaoxiang781216** [comment]
@acassis LGTM.

**Comment by leomar95** [comment]
@acassis I believe it's okay :)

**Comment by acassis** [comment]
Thank you guys!

**Pull request merged**